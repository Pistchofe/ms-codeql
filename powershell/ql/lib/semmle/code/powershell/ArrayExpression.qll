import powershell

class ArrayExpr extends @array_expression, Expr {
  override SourceLocation getLocation() { array_expression_location(this, result) }

  StmtBlock getStmtBlock() { array_expression(this, result) }

  /**
   * Gets the i'th element of this `ArrayExpr`, if this can be determined statically.
   *
   * See `getStmtBlock` when the array elements are not known statically.
   */
  Expr getElement(int i) {
    result =
      unique( | | this.getStmtBlock().getAStmt()).(CmdExpr).getExpr().(ArrayLiteral).getElement(i)
  }

  /**
   * Gets an element of this `ArrayExpr`, if this can be determined statically.
   *
   * See `getStmtBlock` when the array elements are not known statically.
   */
  Expr getAnElement() { result = this.getElement(_) }

  override string toString() { result = "@(...)" }
}
