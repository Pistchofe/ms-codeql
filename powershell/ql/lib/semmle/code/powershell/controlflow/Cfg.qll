/** Provides classes representing the control flow graph. */

import ControlFlowGraph
import CfgNodes as CfgNodes
import BasicBlocks
