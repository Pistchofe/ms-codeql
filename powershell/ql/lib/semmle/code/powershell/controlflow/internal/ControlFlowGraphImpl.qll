/**
 * Provides an implementation for constructing control-flow graphs (CFGs) from
 * abstract syntax trees (ASTs), using the shared library from `codeql.controlflow.Cfg`.
 */

private import powershell
private import codeql.controlflow.Cfg as CfgShared
private import semmle.code.powershell.controlflow.ControlFlowGraph
private import Completion

private module CfgInput implements CfgShared::InputSig<Location> {
  private import ControlFlowGraphImpl as Impl
  private import Completion as Comp
  private import Splitting as Splitting
  private import semmle.code.powershell.Cfg as Cfg

  class Completion = Comp::Completion;

  predicate completionIsNormal(Completion c) { c instanceof Comp::NormalCompletion }

  predicate completionIsSimple(Completion c) { c instanceof Comp::SimpleCompletion }

  predicate completionIsValidFor(Completion c, Ast e) { c.isValidFor(e) }

  class AstNode = Ast;

  class CfgScope = Cfg::CfgScope;

  CfgScope getCfgScope(Ast n) { result = Impl::getCfgScope(n) }

  predicate scopeFirst(CfgScope scope, Ast first) { scope.(Impl::CfgScope).entry(first) }

  predicate scopeLast(CfgScope scope, Ast last, Completion c) {
    scope.(Impl::CfgScope).exit(last, c)
  }

  class SplitKindBase = Splitting::TSplitKind;

  class Split = Splitting::Split;

  class SuccessorType = Cfg::SuccessorType;

  SuccessorType getAMatchingSuccessorType(Completion c) { result = c.getAMatchingSuccessorType() }

  predicate successorTypeIsSimple(SuccessorType t) {
    t instanceof Cfg::SuccessorTypes::NormalSuccessor
  }

  predicate successorTypeIsCondition(SuccessorType t) {
    t instanceof Cfg::SuccessorTypes::ConditionalSuccessor
  }

  predicate isAbnormalExitType(SuccessorType t) {
    t instanceof Cfg::SuccessorTypes::RaiseSuccessor or
    t instanceof Cfg::SuccessorTypes::ExitSuccessor
  }
}

private import CfgInput
import CfgShared::Make<Location, CfgInput>

class CfgScope extends Scope {
  predicate entry(Ast first) { first(this, first) }

  predicate exit(Ast last, Completion c) { last(this, last, c) }
}

/** Holds if `first` is first executed when entering `scope`. */
pragma[nomagic]
predicate succEntry(CfgScope scope, Ast first) { scope.entry(first) }

/** Holds if `last` with completion `c` can exit `scope`. */
pragma[nomagic]
predicate succExit(CfgScope scope, Ast last, Completion c) { scope.exit(last, c) }

/** Defines the CFG by dispatch on the various AST types. */
module Trees {
  // TODO
}

private import Scope

cached
private CfgScope getCfgScopeImpl(Ast n) { result = scopeOf(n) }

/** Gets the CFG scope of node `n`. */
pragma[inline]
CfgScope getCfgScope(Ast n) {
  exists(Ast n0 |
    pragma[only_bind_into](n0) = n and
    pragma[only_bind_into](result) = getCfgScopeImpl(n0)
  )
}

cached
private module Cached {
  cached
  newtype TSuccessorType =
    TSuccessorSuccessor() or
    TBooleanSuccessor(boolean b) { b in [false, true] } or
    TReturnSuccessor() or
    TBreakSuccessor() or
    TRaiseSuccessor() or
    TExitSuccessor()
}

import Cached
