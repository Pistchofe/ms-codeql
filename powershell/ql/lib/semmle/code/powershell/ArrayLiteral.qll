import powershell

class ArrayLiteral extends @array_literal, Expr {
  override SourceLocation getLocation() { array_literal_location(this, result) }

  Expr getElement(int index) { array_literal_element(this, index, result) }

  Expr getAnElement() { array_literal_element(this, _, result) }

  override string toString() { result = "...,..." }
}
