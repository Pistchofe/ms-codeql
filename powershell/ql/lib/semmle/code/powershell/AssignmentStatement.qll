import powershell

class AssignmentStatement extends @assignment_statement instanceof Statement
{
    SourceLocation getLocation()
    {
        assignment_statement_location(this, result)
    }

    int getKind()
    {
        assignment_statement(this, result, _, _)
    }

    Expression getLeftHandSide()
    {
        assignment_statement(this, _, result, _)
    }

    Statement getRightHandSide()
    {
        assignment_statement(this, _, _, result)
    }

    string toString(){
        result = "AssignmentStatement at: " + this.getLocation().toString()
    }
}