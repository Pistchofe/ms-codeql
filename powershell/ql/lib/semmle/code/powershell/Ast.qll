import powershell

class Ast extends @ast {
  string toString() { none() }

  Ast getParent() { parent(this, result) } // TODO: Flip parent and child in relation in the extractor

  Location getLocation() { none() }
}
