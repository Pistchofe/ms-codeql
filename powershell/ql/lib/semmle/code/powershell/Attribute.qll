import powershell

class Attribute extends @attribute, AttributeBase {
  override string toString() { result = this.getName() }

  override SourceLocation getLocation() { attribute_location(this, result) }

  string getName() { attribute(this, result, _, _) }

  int getNumNamedArguments() { attribute(this, _, result, _) }

  int getNumPositionalArguments() { attribute(this, _, _, result) }

  NamedAttributeArgument getNamedArgument(int i) { attribute_named_argument(this, i, result) }

  NamedAttributeArgument getANamedArgument() { result = this.getNamedArgument(_) }

  int getNumberOfArguments() { result = count(this.getAPositionalArgument()) }

  Expr getPositionalArgument(int i) { attribute_positional_argument(this, i, result) }

  Expr getAPositionalArgument() { result = this.getPositionalArgument(_) }

  int getNumberOfPositionalArguments() { result = count(this.getAPositionalArgument()) }
}
