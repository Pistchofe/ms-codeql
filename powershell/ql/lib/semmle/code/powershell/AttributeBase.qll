import powershell

class AttributeBase extends @attribute_base, Ast { }
