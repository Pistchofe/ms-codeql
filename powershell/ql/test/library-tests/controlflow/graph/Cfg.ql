import semmle.code.powershell.Cfg
import semmle.code.powershell.controlflow.internal.ControlFlowGraphImpl::TestOutput<CfgNode>
