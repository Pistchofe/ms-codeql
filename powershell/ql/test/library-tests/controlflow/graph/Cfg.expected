global.ps1:
#    1| {...}
#-----|  -> c

#    1| enter global.ps1
#-----|  -> global.ps1

#    1| exit global.ps1

#    1| exit global.ps1 (normal)
#-----|  -> exit global.ps1

#    1| global.ps1
#-----|  -> a

#    2| a
#-----|  -> 1

#    2| ...=...
#-----|  -> b

#    2| 1
#-----|  -> ...=...

#    2| 1
#-----|  -> 1

#    3| b
#-----|  -> 2

#    3| ...=...
#-----|  -> {...}

#    3| 2
#-----|  -> ...=...

#    3| 2
#-----|  -> 2

#    5| {...}
#-----|  -> exit global.ps1 (normal)

#    6| c
#-----|  -> ...+...

#    6| ...=...
#-----|  -> {...}

#    6| a
#-----|  -> b

#    6| ...+...
#-----|  -> ...=...

#    6| ...+...
#-----|  -> a

#    6| b
#-----|  -> ...+...
